fl4health/utils/functions.py

```diff
@@ -41,3 +45,65 @@ def backward(ctx: torch.Any, grad_output: torch.Tensor) -> torch.Tensor:  # type
 
 def sigmoid_inverse(x: torch.Tensor) -> torch.Tensor:
     return -torch.log(1 / x - 1)
+
+
+def select_random_element(array: np.ndarray) -> float:
+    """
+    Helper function that simply selects a random element from the provided array using np.random.choice.
+
+    Args:
+        array (np.ndarray): Array from which a single element is randomly selected
+
+    Returns:
+        float: randomly selected element value.
+    """
+    indices = tuple(np.random.choice(size) for size in array.shape)
+    return array[indices]
+
+
+def pseudo_sort_function(client_result: Tuple[ClientProxy, NDArrays, int]) -> float:
+    """
+    This function provides the "score" that is used to sort a list of Tuple[ClientProxy, NDArrays, int]. We take a
+    random selection of the array elements and add the integer to them to come up with a score for sorting. Note that
+    the underlying numpy arrays in NDArrays may not all be of numerical type. So we limit to selecting elements from
+    arrays of floats.
+
+    Args:
+        client_result (Tuple[ClientProxy, NDArrays, int]]): Elements to use to determine the score.
+
+    Returns:
+        float: Sum of a random selection of a single element of each array in the NDArrays and the int of the
+        tuple
+    """
+    _, client_arrays, sample_count = client_result
+    random_params = [
+        select_random_element(array) for array in client_arrays if np.issubdtype(array.dtype, np.floating)
+    ]
+    return np.sum(random_params) + sample_count
+
+
+def decode_and_pseudo_sort_results(
+    results: List[Tuple[ClientProxy, FitRes]]
+) -> List[Tuple[ClientProxy, NDArrays, int]]:
+    """
+    This function is used to convert the results of client training into NDArrays and to apply a pseudo sort
+    based on a random selection of the elements in the weights. As long as the numpy seed has been set on the server
```

Comment:
As per our conversation, I'll move to just sampling the first element (zeroeth position) rather than a random on in each layer. I think this should be sufficient to pin the sum order. Will conduct a final test of this tomorrow when the cluster comes back up.

Reply:
I understand John’s point, but this approach could increase the likelihood of collisions, where different model weights might result in the same sum. That said, considering the pros and cons of both options, I would also lean toward sampling only the first element. This way, we avoid introducing extra randomness, even without a set seed.

Reply:
Noted. There is a bit of a risk here, but I think its not large compared to the current implementation. So we'll go with it for now 🙂

Reply:
As an update: Conducted reproducibility test on the cluster and we see determinism with taking the zeroeth elements.

Reply:
Just want to double check, so here in parameters we sum up first element of each Layer tensor, or we only summing up first element of all model weights? Because if it is the second one we would have issues if first layer is frozen for example, if first one everything look great for me.

Reply:
This will iterate through all NDarrays passed to the server for aggregation. In general, these represent are tensor used by the model. So it will be the sum of the first element in each tensor. For example, if there is a linear layer with bias, there will likely be two tensors here (one for the linear transform and one for the bias). Each tensor will have the first element extracted and then they will be summed.

Reply:
I see, then it should be the first scenario and good to go.